### ISSUE TITLE
[contracts] Introduce immutable variables to WASM contracts

### ISSUE THREAD

**Issue by agryaznov** [paritytech/substrate#13448]
Solidity has the concept of [immutable variables](https://docs.soliditylang.org/en/latest/contracts.html#constant-and-immutable-state-variables), which can be set once in contract&rsquo;s constructor and guaranteed to stay unchanged afterwards. _OpenZeppelin_ has recently recommended *ink!* to introduce this feature too, to provide contract developers with more flexibility in designing and building secure and efficient smart contracts. In this issue we&rsquo;ll try to think about how we can introduce the similar concept into *ink!* and Pallet Contracts.

WebAssembly has indeed the similar construct - [immutable global](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-global). Its initializer lives in the Globals section of the WASM module, and uses a constant initializing expression.

The constructor of WASM contracts in the *pallet_contracts* is the `deploy` exported function. This function is called once at the time of contract instantiation on-chain. Contract instantiation is done after the re-instrumentation and uploading the contract WASM blob on-chain, and it in general can be done in a separate (from uploading) extrinsic, when deploying a new contract instance of a code already stored on-chain. This approach allows to save gas and storage deposit for the caller account. Changing the *Global* section of the module on the step of contract instantiation would require its code altering and re-uploading on chain, implying an additional costs for contract authors.

Another approach could be to use an imported global in the contract module *(note: currently importing globals to WASM contracts [is forbidden](https://github.com/paritytech/substrate/blob/56de870f943679fede8443d06d3aa96a9f7321d6/frame/contracts/src/wasm/prepare.rs#L323))*. Pallet contracts could define the global and initialize it in the &ldquo;env&rdquo; module [during the instantiation](https://github.com/paritytech/substrate/blob/56de870f943679fede8443d06d3aa96a9f7321d6/frame/contracts/src/wasm/mod.rs#L242) of the module to the WASM engine. For that it would need to somehow get the initialing expression from the constructor (`deploy` exported function of a module, which is to be designed by contract developer in `#[ink(constructor)]`). We can e.g. expose a host function `init_immutable_global` to the contracts API for that purpose on the pallet side.

But the core problem here is that we instantiate the module first to the WASM engine, and only after that we call the exported `deploy` function. My gut feeling it that probably we&rsquo;d have to use *WebAssembly* [`start`](<https://webassembly.github.io/spec/core/syntax/modules.html#syntax-start>) function for that, which is by design intended for initializing the state of a module (and is currently [forbidden](https://github.com/paritytech/substrate/blob/56de870f943679fede8443d06d3aa96a9f7321d6/frame/contracts/src/wasm/mod.rs#L246) in WASM contracts as well).

@athei  @cmichi @ascjones @HCastano wdyt?

**Comment by athei** [comment]
Immutable variables is a programming language feature. It is in the case of solidity and it should be in our case, too. I see no reason to bake this complexity into `pallet-contracts`. As a rule of thumb: If a feature can be implemented in user space this is where it should be implemented. Adding complexity (and policy) to privileged code should only be done with a very good rationale.

In this case I don't see a bid advantage in moving this logic to pallet-contracts: ink! can solve this problem by flagging certain fields as read only: We remove direct field access for the root storage struct and generate accessor functions (we leave out the setter for read only variables).

**Comment by HCastano** [comment]
I agree with Alex here. Can you move this into the ink! repo and spec it out based on ink!'s perspective instead?

**Comment by xlc** [comment]
Anything needs to be done other than just use Rust's const ?
